

<h2>Analyse enrichments</h2>

<br><br>

<hr>
<hr>


<h4>Aim</h4>
<p align="justify">In this tutorial you will analyse ChIP-Seq signals (enrichment of ChIP-Seq reads) in a subset of mouse promoters bound by specific transcription factors. Topics:
<li><a href="#import">Importing the required files</a></li>
<li><a href="#overlaps">Define promoters bound by Myc and Yap transcription fators</a></li>
<li><a href="#resize">Resizing regions of interest</a></li>
<li><a href="#associateEnrichments">Associating enrichment files to a region of interest</a></li>
<li><a href="#analogicHeat">Analysing enrichments with interactive heatmaps</a></li>
<li><a href="#extractROI">Extracting regions of interest based on particular enrichment patterns</a></li>
<li><a href="#correlations">Finding correlations between enrichments</a></li>
</p>


<hr>
<hr>

<h4>Requirements</h4>
<p align="justify">
 <ul>
  Session file "tutorial_ROI.rds" and enrichment files in bigWig format (yapmyc_ChIPSeq_YAP.bw, yapmyc_ChIPSeq_Pol2.bw, yapmyc_ChIPSeq_MYC.bw, yapmyc_ChIPSeq_H3K27Ac.bw, yapmyc_ChIPSeq_H3K4me3.bw) available at <a href=https://drive.google.com/drive/folders/1BrYso234fcpL2wir9UhMR97iIw9_4fRx?usp=sharing> this link</a>

</ul> 
</p>





<h4><a id="import">1) Opening the session file and importing enrichments</a></h4>
<p align="justify">To import the session file generated in the previous tutorial, click "Load session file" button on the side menu (fig. 1). Then choose the file "tutorial_ROI.rds" from the folder in which you downloaded the tutorial files. All the required ROIs will be imported in ChroKit.
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig1.png' width="15%" alt='missing' />
    <figcaption>Figure 1</figcaption>
</figure>

To import the enrichment files, click "Enrichment files" on the side menu ("Import data" section, fig. 2a), then "Choose one or more files from filesystem" (fig. 2b) and press "Select a file" button. Navigate through the filesystem and choose the yapmyc_ChIPSeq_YAP.bw. Repeat the operation for all the other bigWig files (i.e. yapmyc_ChIPSeq_Pol2.bw, yapmyc_ChIPSeq_MYC.bw, yapmyc_ChIPSeq_H3K27Ac.bw, yapmyc_ChIPSeq_H3K4me3.bw). The path for all the enrichment files is shown in the window in the middle (fig. 2c).
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig2.png' width="70%" alt='missing' />
    <figcaption>Figure 2</figcaption>
</figure>
</p>



<h4><a id="overlaps">2) Identifying promoters bound by the transcription factors Myc and Yap</a></h4>
<p align="justify">
On side menu, click "ROI management" ("Work with data" section, fig. 3a) and select the "Overlaps" tab (fig. 3b). Select "promoters" as a reference ROI (fig. 3c) and Myc and Yap peaks as contrast ROIs (yapmyc_ChIPSeq_MYC.xls and yapmyc_ChIPSeq_YAP.xls ROIs, fig. 3d); choose "Intersection of the selected ROIs" below, since we want the promoters to overlap with both Yap and Myc (fig. 3e). Finally, type the name of the new ROI in the dedicated space ("promoters_bound_yap_myc") and click the "Build the ROI!" button (fig. 3f).
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig3.png' width="70%" alt='missing' />
    <figcaption>Figure 3</figcaption>
</figure>
After this step, you will obtain 9433 mouse promoters bound by both Myc and Yap in the tet-Yap/Myc transgenic.
</p>


<h4><a id="resize">3) Resizing promoters</a></h4>
<p align="justify">
For a wider view of the enrichments around TSS, you need to increase the width of the promoters (by default defined as a 600 bp window around TSS). On the side menu, click "ROI management" ("Work with data" section, fig. 4a) and select the "Modify ROIs" tab (fig. 4b): six windows will appear, each containing a different function for modifying ROIs. In the "Resize" window, choose the "promoters_bound_yap_myc" ROI created in step 2 (fig. 4c), select 2500 base pairs upstream and downstream the middle of the ranges (fig. 4d) and type the name of the new ROI ("promoters_bound_resized", fig. 4e); finally, click "Create ROI" button (fig. 4f). 
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig4.png' width="40%" alt='missing' />
    <figcaption>Figure 4</figcaption>
</figure>
The new ROI called "promoters_bound_resized" will be added to the ROIs list.
</p>


<h4><a id="associateEnrichments">4) Associating enrichments (bigWig) to promoters</a></h4>
<p align="justify">
Before analysing enrichments, you must associate enrichment files to one or more ROIs (i.e. finding the reads coverage for each position in the genomic ranges of the ROI). As an example, you will associate the enrichment files (in bigWig format) of ChIP-Seq experiments in tet-Yap/Myc transgenic mouse. On the side menu, click "ROI management" ("Work with data" section, fig. 5a) and select the "promoters_bound_resized" ROI created in the step before (fig. 5b). Then choose all the enrichments available (fig. 5c). Note: unless a large amount of free RAM is available on your machine, be sure to use only one CPU core (fig. 5d). Finally, click the "Associate!" button. The loading gif will appear on the left (fig. 5e), indicating that an operation is ongoing. WARNING: Do not carry out other operations while loading, but wait until finished.
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig5.png' width="60%" alt='missing' />
    <figcaption>Figure 5</figcaption>
</figure>
The association will take few or several seconds (depending on the computational power of the machine in use); once completed, the subset of promoters will be available for the enrichment analyses.
</p>


<h4>5) Generate enrichment-based heatmaps</h4>
<p align="justify">
	<b><a id="analogicHeat">Generate the heatmap</a></b>
	<br>
Click "Genomics" on the side menu ("Work with data" section, fig. 6a), and select the "Enrichment-based heatmap" tab (fig. 6b). Here, select the "promoters_bound_resized" ROI created in step 3 (fig. 6c) and all the enrichments associated with it (fig. 6d). Select the "clustering" option and all the enrichments as input for the algorithm (fig. 6e), leaving the default clustering options. Finally, click the "Update plot" button and the heatmap will appear on the right.
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig6.png' width="70%" alt='missing' />
    <figcaption>Figure 6</figcaption>
</figure>

To adjust additional Heatmap parameters, go to the "Advanced" tab (fig. 7a). If you want to display all the genomic ranges, select a random sample of 7000 ranges (or a number higher than the number of genomic ranges of the master ROI, in this case 6900) (fig. 7b). Click "Update plot" again to build the heatmap showing all the genomic ranges in the ROI; then set the "quantile threshold" to "individual" (fig. 7c) so that signals can be compared among different ChIP-Seq experiments (this sets the same color scale for all the signals displayed). To find the optimal color saturation, modify the threshold using the slide-bar (fig. 7d). Select "custom colors" to modify the colors of the heatmap (fig. 7e). Finally, click the "Get PDF" button (fig. 7f) to download the editable PDF file of the heatmap displayed.
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig7.png' width="70%" alt='missing' />
    <figcaption>Figure 7</figcaption>
</figure>
Note: The maximum number of available genomic ranges for the heatmap (6900) is lower compared to those in the promoters_bound_yap_tead ROI (9433). This is because if alternative promoters of the same genes are identical, they are collapsed into a single promoter.
<br><br>
<b><a id="extractROI">Extract regions of interest from the heatmap</a></b>
<br>
Clusters are depicted as color bars on the left of the heatmap (fig. 8a). We are interested in the first and fourth clusters, which are enriched in H3K4me3 upstream and downstream TSS, respectively. To extract these genomic ranges, click the corresponding bar (fig. 8b), type the name of the new ROI ("promoters_upstream_enrichment", fig. 8c) and click "Import ROI" (fig. 8d): a new ROI of 1753 genomic ranges will be generated.  
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig8.png' width="70%" alt='missing' />
    <figcaption>Figure 8</figcaption>
</figure>
Repeat the same operation for the cluster number four, giving the name "promoters_downstream_enrichment" to the new ROI. 

To show the enrichments of RNAPol2 in different clusters as profiles and boxplots, select the area of interest in the heatmap by pointing and dragging with the mouse/trackpad (choose the Pol2 column, fig. 9a). By dragging this area, for example moving the window to the Myc column, profiles (fig. 9b) and boxplots (fig. 9c) are immediately updated in the windows below.

<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig9.png' width="70%" alt='missing' />
    <figcaption>Figure 9</figcaption>
</figure>
Genomic ranges of the selected area can also be extracted as shown above for the clusters (fig. 8c and 8d).
To find if promoters with a particular H3K4me3 enrichment pattern are associated with particular gene programs, carry out a gene ontology analysis following the ROI tutorial (step number 6), choosing the "promoters_upstream_enrichment" and "promoters_downstream_enrichment" ROIs.
</p>


<h4><a id="correlations">6) Evaluating correlations between enrichments</a></h4>
<p align="justify">
To understand the relationship between enrichments associated to the promoters bound by the transcription factors Myc and Yap, click "Genomics" on the side menu ("Work with data" section, fig. 10a) and select "Enrichments in ROIs" tab (fig. 10b). Select the "promoters_bound_resized" ROI from the menu (fig. 10c) and yapmyc_ChIPSeq_YAP.bw, yapmyc_ChIPSeq_Pol2.bw and yapmyc_ChIPSeq_MYC.bw enrichments (fig. 10d). Check the "log2" button (fig. 10e) for the log transformation of data and press the "Update plot" button (fig. 10f). Different plots will be generated in the windows on the right, including signal profiles, boxplots and a correlation heatmap (fig. 10g). To view the scatterplot of the correlation between Myc and Pol2, click the corresponding square of the heatmap (fig. 10h) and the scatterplot will appear on the right (fig. 10i).
<figure>
    <img src='../tutorial_figures/tutorial_enrichments_fig10.png' width="70%" alt='missing' />
    <figcaption>Figure 10</figcaption>
</figure>
</p>



